### ISSUE TITLE
Let FTW link to a shared SQLite SO

### ISSUE THREAD

**Issue by DavidStaab** [featherweight/ftw-kernel#1]
FTW builds SQLite into its own .so. This works if all actors in a system are using FTW to access SQLite, but many designs (including almost all of ours) involve components from multiple languages and binaries sharing access to a SQLite database. SQLite's documentation explicitly advises the use of a single, common copy of sqlite3.so: https://www.sqlite.org/howtocorrupt.html (see "2.2.1 Multiple copies of SQLite linked into the same application").

It would be more friendly to polyglot applications if FTW linked to SQLite3 instead of composing it.

Additionally, SQLite3 is revised and republished on opkg and ipkg repositories regularly. An FTW update will be required to incorporate every fix or addition to SQLite that's published by their team.

**Comment by wirebirdlabs** [comment]
Dynamic vs. static linking -- big conversation :-)

Briefly, static linking is used for all libraries in `ftw-kernel` in order to provide a well-defined, curated experience. Dynamic linking could provide a more general/flexible solution, yet while bringing on the associated problems of dependency management.

In regards to a few of your points:

> SQLite3 is revised and republished on opkg and ipkg repositories regularly. An FTW update will be required to incorporate every fix or addition to SQLite that's published by their team

The inverse might hold as well, depending on which repository you're using. For instance, ftw-kernel currently uses 3.9.1 (2015-10-16), whereas http://download.ni.com/ni-linux-rt/feeds/2015sp1/arm/ipk/cortexa9-vfpv3/ is at 3.8.6 (2014-08-15, 14months older).

We're certainly not arguing that FTW is superior here by being the most up-to-date -- it's almost certainly behind some other feed you can find. The argument here though, is that when you *need* to upgrade for an urgent project requirement, `ftw-kernel` is more accessible for self-help, rather than relying upon a third-party package maintainer is able to push an update (which, is likely a task you'd be able to do as well, since that project is likely open source as well, but then there's the associated cost of tooling up with that new toolchain ...).

> polyglot applications if FTW linked to SQLite3 instead of composing it.

Can you give a specific example? It totally makes sense what you're asking for here, and I can contrive a few examples where it seems useful, but I'm curious the relative benefit.

Basically -- nothing you're asking for is technically invalid or undesirable -- it's just that the features here are inherently coupled with a different type of problem, and that type of problem is one FTW has been more concerned with avoiding.

**Comment by DavidStaab** [comment]
> Can you give a specific example [of a polyglot application]?

A JavaScript or Python app that accesses a common database with the LV app running on the same cRIO. Specific uses for the database vary heavily, and I'd rather not constrain the discussion of value to any poorly chosen ones I cite here.

If FTW's purpose is to provide a one-stop tool library for LV development, I think that should be made more clear in the project's high-level documentation. If the goal is to enhance LV's interoperability with the rest of the world, bundling your own distro of SQLite seems to undermine that goal.

**Comment by wirebirdlabs** [comment]
This makes sense. What you're saying are the specific reasons that `sqlite` could be treated differently from the other libraries amalgamated into the single `libftw`, which don't have the same technical reasons for dynamic linking.

That said ... it would take some technical investigation to see if the problem cited in 2.2.1 at https://www.sqlite.org/howtocorrupt.html still even holds -- it's *possible* that such global state held by the library has since been refactored away, or fully mitigated by one of the more rigorous synchronization PRAGMA settings. Assuming that was the case -- does the value proposition of dynamic loading still hold? (this is a rhetorical question; I don't have an answer.

Anyway, if we did develop the capability of dynamically-loading `sqlite`, it would probably be a make-time opt-in option, where the default is to statically link the curated version that comes with FTW.

Again, all certainly makes sense, but is a substantial amount of work for what might be a small fraction of usage scenarios.
